base/compiler/tfuncs.jl

```diff
@@ -1122,13 +1129,32 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
             #        ai = ai.body
             #    end
             else
+                # Is this the second parameter to a NamedTuple?
+                if isnamedtuple && isa(ua, UnionAll) && uw.parameters[2] === ua.var
+                    # If the names are known, keep the upper bound (which we know is
+                    # at most Tuple), but otherwise widen to Tuple.
+                    # This is a widening heuristic to avoid keeping type information
+                    # that's unlikely to be useful.
+                    if uw.parameters[1] isa Tuple || (i == 2 && tparams[1] isa Tuple)
+                        ub = typeintersect(ub, Tuple)
+                    else
+                        ub = Tuple
+                    end
+                else
+                    ub = Any
+                end
```

Comment:
```suggestion
                    if !(uw.parameters[1] isa Tuple || (i == 2 && tparams[1] isa Tuple))
                        ub = Any
                    end
                else
                    ub = Any
                end
```

Reply:
Ah, yes. The `Tuple` bound should be valid only if `ai_w <: Type` above holds, i.e. if whatever we're putting in there is not a typevar.

Reply:
I think not even then?
```
julia> NamedTuple{<:Any, Val{T}} where T
NamedTuple{var"#s1",Val{T}} where var"#s1" where T
```

Reply:
Ouch.